packages/cozy-realtime/src/CozyRealtime.js

```diff
@@ -1,360 +1,534 @@
+      this.websocket.onopen = null
+      this.websocket.onclose = null
+      try {
+        this.websocket.close()
+      } catch (e) {
+        // void
+      }
+      this.websocket = null
+    } else {
+      logger.warn(
+        'Trying to reject a websocket that is not attached. This should not happen'
+      )
+    }
   }
 
-  if (msg) {
-    logger.error(msg)
-    throw new Error(msg)
+  /**
+   * Close the Realtime
+   * @private
+   */
+  stop() {
+    if (this.isStarted) {
+      logger.info('stopped')
+      this.unsubscribeGlobalEvents()
+      this.unsubscribeClientEvents()
+      if (this.hasWebSocket()) {
+        this.rejectWebSocket()
+      }
+      this.isStarted = false
+      if (!this.subscriptions.isEmpty()) {
+        this.unsubscribeAll()
+      }
+      this.emit('close')
+    }
   }
-}
 
-/**
- * Return websocket url from cozyClient
- *
- * @return {String}  WebSocket url
- */
-export const getWebSocketUrl = cozyClient => {
-  const isSecureURL = url => !!url.match(`^(https:/{2})`)
-
-  const url = cozyClient.stackClient.uri
-  const protocol = isSecureURL(url) ? 'wss:' : 'ws:'
-  const host = new URL(url).host
-
-  return `${protocol}//${host}/realtime/`
-}
-
-/**
- * Return token from cozyClient
- *
- * @return {String}  token
- */
-export const getWebSocketToken = cozyClient =>
-  cozyClient.stackClient.token.accessToken || cozyClient.stackClient.token.token
-
-/**
- * CozyRealtime class
- *
- * @class
- */
-class CozyRealtime {
   /**
-   * A cozy client
-   *
-   * @type {CozyClient}
+   * Authenticate to the websocket
+   * @private
    */
-  _cozyClient = null
+  authenticate() {
+    if (this.isWebSocketOpen()) {
+      const token = getToken(this.client)
+      logger.debug('authenticate with', token)
+      this.sendWebSocketMessage('AUTH', token)
+      this.websocket.authenticated = true
+    } else {
+      logger.warn(
+        'Trying to authenticate to a non-opened websocket. This should not happen.',
+        this.websocket,
+        this.websocket && this.websocket.readyState
+      )
+    }
+  }
 
   /**
-   * A Socket class
+   * Send a message through the websocket
    *
-   * @type {Socket}
+   * @private
+   * @param {string} method - like 'AUTH', 'SUBSCRIBE' or 'UNSUBSCRIBE'
+   * @param {object} payload - message to be sent
    */
-  _socket = null
+  sendWebSocketMessage(method, payload) {
+    const message = JSON.stringify({ method, payload })
+    this.websocket.send(message)
+  }
 
   /**
-   * Delay (ms) to retry socket connection
+   * Send a realtime notification to the server
    *
-   * @type {Interger}
+   * @param {string} doctype
+   * @param {string} id
+   * @param {object} data
    */
-  _retryDelay = 1000
+  async sendNotification(doctype, id, data) {
+    return this.client
+      .getStackClient()
+      .fetchJSON('POST', `/realtime/${doctype}/${id}`, {
+        data
+      })
+  }
+  send(...args) {
+    return this.sendNotification(...args)
+  }
```

Comment:
Why this abstraction? Can't we have a direct call to sendNotification? Or renaming sendNotification to send?

Reply:
Because the previous class had a direct `send` method. I found it misleading (I would expect something to send a message through the websocket, but it does something else) so I made a new name but kept the old one for compatibility.

Reply:
Could we have a deprecation warning here ?